Core/Src/modo.c

```diff
@@ -0,0 +1,49 @@
+/*
+ * modo.c
+ *
+ *  Created on: Jul 13, 2021
+ *      Author: Felipe Telles
+ */
+
+
+#include "modo.h"
+#include "cmsis_os.h"
+#include "global_variables.h"
+#include "global_instances.h"
+
+
+
+void seleciona_modo(void *argument) {
+
+	for(;;) {
+
+		#ifdef DEBUG_ECU
+		extern void brkpt();
+		brkpt();
+		#endif
+
```

Comment:
Vc vai precisar verificar se está em RTD aqui. Se tiver, n faz sentido executar esse código, e nem pode. Imagina a situação:

- entrou em RTD
- A pessoa apertou o botão de mudar de modo (não vai funcionar, pq tá em RTD, mas o código aq vai continuar a execução mas vai ficar travado ali no semáforo)
- Assim q sair do RTD, o modo tbm vai mudar (lembra q ele ficou travado no semáforo ali embaixo, e agr foi destravado e terminou a execução)

esse comportamento é inesperado, então vc precisa fazer essa verificação de RTD aqui.

Pra fazer a verificação, vc pode dar um flagGet e verificar se tá em RTD. Se tiver, dá um delay de 1s e volta (importante dar o delay, se n essa tarefa vai fica consumindo recurso)

Reply:
Posso fazer esse verificação extra pra redundancia. Mas no código como está agora isso não gera um comportamento inesperado.

> - A pessoa apertou o botão de mudar de modo (não vai funcionar, pq tá em RTD, mas o código aq vai continuar a execução mas vai ficar travado ali no semáforo)

Sim o modo não irá mudar,  `g_race_mode` vai mudar, mas n importa.

> - Assim q sair do RTD, o modo tbm vai mudar (lembra q ele ficou travado no semáforo ali embaixo, e agr foi destravado e terminou a execução) 

Antes de sair RTD `modo_selecionado` e `g_race_mode` é configurado pra ser igual erro, então quando começa o for do inicio e o semaphore é liberado o máximo que acontece é `modo_selecionado` ser setado para erro novamente.

Não achei necessário fazer a verificação de RTD em modos.c, mas se ainda achar que é necessário posso colocar, não vai dá mt trabalho

Reply:
entendi. Tem algum cenário pra ele sair do modo de erro?

Reply:
Só apertar o botão de mudar o modo (mas só funciona se apertar depois que sair do RTD, antes não adianta, testei apertando antes e sempre vai pra erro).
Essa é uma área bem aberta. Não tem no regulamento nada que precisa sair de RTD, apenas zerar referencia de torque. Então é bem aberto esses erros que saem do RTD, por isso coloquei que só apertar botão já permite trocar de modo. Pensei até em precisar apertar o botão de RTD e modo ao msm tempo, mas deixei quieto.
Talvez seja interessante criar uma combinação de botões pra reiniciar a ECU ou só tirar de RTD.